nextstrain/cli/remote/nextstrain_dot_org.py

```diff
@@ -0,0 +1,697 @@
+    A remote Nextstrain dataset, as described by a Charon API response,
+    extended for the nextstrain.org RESTful API.
+    """
+    def __init__(self, api_item):
+        super().__init__(api_item)
+
+        default_sidecars = ["root-sequence", "tip-frequencies"]
+
+        self.subresources = [
+            SubResource("application/vnd.nextstrain.dataset.main+json", ".json", primary = True),
+
+            # XXX TODO: The "sidecars" field in the /charon/getAvailable API
+            # response doesn't actually exist yet and its use here is
+            # prospective.
+            #
+            # I plan to extend the /charon/getAvailable API endpoint (or maybe
+            # switch to a new endpoint) in the future to include the "sidecars"
+            # field listing the available sidecars for each dataset, so that
+            # this code only has to try to fetch what is reported to exist.
+            # More than just reducing requests, the primary upshot is looser
+            # coupling by avoiding the need to update the hardcoded list of
+            # sidecars here and get people to upgrade their installed version
+            # of this CLI if we add a new sidecar in the future.  Other API
+            # clients would also likely benefit.
+            #
+            #   -trs, 18 August 2021
+            #
+            *[SubResource(f"application/vnd.nextstrain.dataset.{type}+json", ".json")
+                for type in api_item.get("sidecars", default_sidecars)],
+        ]
+
+
+class Narrative(Resource):
+    """
+    A remote Nextstrain narrative, as described by a Charon API response,
+    extended for the nextstrain.org RESTful API.
+    """
+    subresources = [
+        SubResource("text/vnd.nextstrain.narrative+markdown", ".md", primary = True),
+    ]
+
+
+def upload(url: urllib.parse.ParseResult, local_files: List[Path]) -> Iterable[Tuple[Path, str]]:
+    """
+    Upload the *local_files* to the given nextstrain.org *url*.
+    """
+    path = remote_path(url)
+
+    # XXX TODO: Consider changing the RemoteModule interface for upload() and
+    # organizing files in the upload command's run() instead of here.  This
+    # would make most sense as part of a larger UI rework of the `remote`
+    # commands to focus on datasets/narratives rather than local/remote files.
+    #
+    #   -trs, 22 Sept 2021
+    datasets, narratives, unknowns = organize_files(local_files)
+
+    def file_list(files: Iterable):
+        return indent("\n".join(map(str, files)), "    ")
+
+    if unknowns:
+        # XXX TODO: Handle group-logo.png and group-overview.md by making
+        # appropriate API requests to the group endpoint (which doesn't yet
+        # exist).
+        #   -trs, 23 Sept 2021
+        raise UserError(f"""
+            Only datasets (v2) and narratives are currently supported for
+            upload to nextstrain.org, but other files were given:
+
+            {{files}}
+            """, files = file_list(unknowns))
```

Comment:
Uploading unexpected files technically wouldn't "break" anything, which is why I thought this is overly cautious. 

I do agree it's not nice to leave a mess for the user to clean up. This makes me think it's insufficient that we only catch this problem sometimes. Maybe we should check this systematically and not just rely on the coincidence that they included an unknown file. 

It would be nice to get confirmation from the user when uploading files. This can be a list of the datasets and narratives or even just a summary of "Please confirm X datasets and X narratives will be uploaded". We can always add a `-f` option to allow users to bypass this confirmation step.

Reply:
It wouldn't "break" things in the sense of "this didn't work", but it breaks things in the sense that there'd be garbage datasets/narratives listed on the Groups page, private info might be leaked, etc. And I think it breaks an expectation that these commands won't let you upload just any old files.

I agree the detection can be more systematic, hence my comments about using our schemas to do detection pre-upload instead of naive filename matching.

Confirmation is a good suggestion; I've updated #149 to describe a `--prompt` (and `--no-prompt`) mode because they're both related to `--dry-run`.